common/core/web/input-processor/src/text/inputProcessor.ts

```diff
@@ -128,19 +129,25 @@ namespace com.keyman.text {
 
               let altEvent = altKey.constructKeyEvent(this.keyboardProcessor, mock, keyEvent.device);
               let alternateBehavior = this.keyboardProcessor.processKeystroke(altEvent, mock);
-              if(alternateBehavior) {
-                // TODO: if alternateBehavior.beep == true, set 'p' to 0.  It's a disallowed key sequence,
-                //       so a user should never have intended to type it.  Should probably renormalize 
-                //       the distribution afterward, though...
-                
+              
+              // If alternateBehavior.beep == true, ignore it.  It's a disallowed key sequence,
+              // so we expect users to never intend their use.
+              if(alternateBehavior && !alternateBehavior.beep) {
                 let transform: Transform = alternateBehavior.transcription.transform;
                 
                 // Ensure that the alternate's token id matches that of the current keystroke, as we only
                 // record the matched rule's context (since they match)
                 transform.id = ruleBehavior.transcription.token;
                 alternates.push({sample: transform, 'p': pair.p});
+                totalMass += pair.p;
               }
             }
+
+            // Renormalizes the distribution, as any error (beep) results
+            // will result in a distribution that doesn't sum to 1 otherwise.
+            alternates.forEach(function(alt) {
+              alt.p /= totalMass;
```

Comment:
I'd really like a div/0 check here -- even if it is unlikely (what about a keyboard layer with no alternate behaviour keys?)

Reply:
If there are no alternate behavior keys, then there are no entries in the array and the loop never iterates.

I've added a bit of code to ensure any `.p` entries are strictly positive, after which we can logically guarantee `totalMass > 0` for this line whenever it is reached.